src/main/java/com/arpnetworking/metrics/mad/sources/TransformingSource.java

```diff
@@ -0,0 +1,490 @@
+                                            metric.getValue(),
+                                            newMetricName,
+                                            mergedMetrics,
+                                            getModifiedDimensions(record.getDimensions(), parsedTags, consumedDimensions, transformation));
+                                }
+                            }
+                            //Having "found" set here means that mapping a metric to an empty list suppresses that metric
+                            found = true;
+                        }
+                    }
+                    if (!found) {
+                        merge(
+                                metric.getValue(),
+                                metricName,
+                                mergedMetrics,
+                                getModifiedDimensions(record.getDimensions(), Collections.emptyMap(), ImmutableList.of(), transformation));
+                    }
+                }
+
+                // Raise the merged record event with this source's observers
+                // NOTE: Do not leak instances of MergingMetric since it is mutable
+                for (final Map.Entry<Key, Map<String, MergingMetric>> entry : mergedMetrics.entrySet()) {
+                    _source.notify(
+                            ThreadLocalBuilder.build(
+                                    DefaultRecord.Builder.class,
+                                    b1 -> b1.setMetrics(
+                                            entry.getValue().entrySet().stream().collect(
+                                                    ImmutableMap.toImmutableMap(
+                                                            Map.Entry::getKey,
+                                                            e -> ThreadLocalBuilder.clone(
+                                                                    e.getValue(),
+                                                                    DefaultMetric.Builder.class))))
+                                            .setId(record.getId())
+                                            .setTime(record.getTime())
+                                            .setAnnotations(record.getAnnotations())
+                                            .setDimensions(entry.getKey().getParameters())));
+                }
+            }
+        }
+
+        private Key getModifiedDimensions(
+                final ImmutableMap<String, String> inputDimensions,
+                final Map<String, String> add,
+                final List<String> remove,
+                final TransformationSet transformation) {
+            final Map<String, String> finalTags = Maps.newHashMap(inputDimensions);
+            // Remove the dimensions that we consumed in the replacement
+            remove.forEach(finalTags::remove);
+            transformation.getRemoveDimensions().forEach(finalTags::remove);
+            transformation.getInjectDimensions().forEach(
```

Comment:
You should document that inject trumps remove and both of these trump anything extracted in the transform and subsequently removed. However, I'm a little concerned about the way this is done. Specifically, the way these can compound. Depending on your feedback on the independent UOWs from each transformation this may be a discussion we should have.

Reply:
That's not how this is done. 1) remove the consumed dimensions, 2) remove the explicitly removed dimensions, 3) inject the explicitly injected dimensions (honoring overwrite), 4) inject the extracted dimensions.  Not sure what you think the order should be.